ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -12,17 +25,332 @@ pub trait DZKPCompatibleField: Field {
+                                // g2=c(1-2e),
+                                Fp61BitPrime::from_bit(*c) * one_minus_two_e,
+                                // g3=a(1-2e),
+                                Fp61BitPrime::from_bit(*a) * one_minus_two_e,
+                                // g4=-1/2(1-2e),
+                                Fp61BitPrime::MINUS_ONE_HALF * one_minus_two_e,
+                            ]
+                        })
+                        .collect::<GenericArray<Fp61BitPrime, InitialRecursionFactor>>(),
+                    // h polynomial
+                    b.iter()
+                        .zip(d.iter())
+                        .zip(f.iter())
+                        .zip(bd.iter())
+                        .flat_map(|(((b, d), f), bd)| {
+                            let one_minus_two_f = Fp61BitPrime::ONE
+                                + Fp61BitPrime::MINUS_TWO * Fp61BitPrime::from_bit(*f);
+                            [
+                                // h1=bd(1-2f),
+                                Fp61BitPrime::from_bit(*bd) * one_minus_two_f,
+                                // h2=d(1-2f),
+                                Fp61BitPrime::from_bit(*d) * one_minus_two_f,
+                                // h3=b(1-2f),
+                                Fp61BitPrime::from_bit(*b) * one_minus_two_f,
+                                // h4=1-2f,
+                                one_minus_two_f,
+                            ]
+                        })
+                        .collect::<GenericArray<Fp61BitPrime, InitialRecursionFactor>>(),
+                )
+            })
+    }
+
+    // Convert allows to convert individual bits from multiplication gates into dzkp compatible field elements
+    // We use the conversion logic from https://eprint.iacr.org/2023/909.pdf
+    //
+    // This function does not use any optimization.
+    //
+    //
+    // Prover and left can compute:
+    // g1=-2ac(1-2e),
+    // g2=c(1-2e),
+    // g3=a(1-2e),
+    // g4=-1/2(1-2e),
+    //
+    // where
+    // (a,c,e) = (xright, yright, xright * yright ⊕ zright ⊕ prssright)
+    fn convert_verifier_left<'a>(
+        x_right: &'a BitArray32,
+        y_right: &'a BitArray32,
+        prss_right: &'a BitArray32,
+        z_right: &'a BitArray32,
+    ) -> impl Iterator<Item = SingleUVPolynomial<Self>> {
+        x_right
+            .chunks(RECURSION_CHUNK_SIZE)
+            .zip(y_right.chunks(RECURSION_CHUNK_SIZE))
+            .zip(prss_right.chunks(RECURSION_CHUNK_SIZE))
+            .zip(z_right.chunks(RECURSION_CHUNK_SIZE))
+            .map(|(((a, c), prss), z)| {
+                // precompute ac = a & c
+                let ac: BitArray<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0> = BitArray::try_from(a)
+                    .unwrap()
+                    & BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(c).unwrap();
+                // e = ac ⊕ zright ⊕ prssright
```

Comment:
This is the first reference to this alternative formulation of `e`. I'd recommend putting the derivation for this up in the comments above with the other one.

Reply:
I have added a comment on top of the function: 
```
// here e is defined as in the paper (since the the verifier does not have access to b,d,f,
// he cannot use the simplified formula for e)
```